src/Kestrel.Transport.Abstractions/Internal/TransportConnection.Features.cs

```diff
@@ -0,0 +1,215 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO.Pipelines;
+using System.Net;
+using System.Text;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Protocols.Features;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions.Internal
+{
+    public partial class TransportConnection : IFeatureCollection,
+                                               IHttpConnectionFeature,
+                                               IConnectionIdFeature,
+                                               IConnectionTransportFeature
+    {
+        private static readonly Type IHttpConnectionFeatureType = typeof(IHttpConnectionFeature);
+        private static readonly Type IConnectionIdFeatureType = typeof(IConnectionIdFeature);
+        private static readonly Type IConnectionTransportFeatureType = typeof(IConnectionTransportFeature);
+        private static readonly Type IConnectionApplicationFeatureType = typeof(IConnectionApplicationFeature);
+
+        private object _currentIHttpConnectionFeature;
+        private object _currentIConnectionIdFeature;
+        private object _currentIConnectionTransportFeature;
+        private object _currentIConnectionApplicationFeature;
+
+        private int _featureRevision;
+
+        private List<KeyValuePair<Type, object>> MaybeExtra;
```

Comment:
The Maybe in MaybeExtra makes is meant to make it abundantly clear that the field can be null. I'm pretty sure it's a haskellism carried over from when @lodejard first code generated our FrameHeaders class. Personally, I like it a lot. Starting a private field name with a capital letter though...